analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            if (!compilationStart.Compilation.ReferencesControllers())
+            {
+                return;
+            }
+            var examinedTypes = new ConcurrentDictionary<ITypeSymbol, bool>();
+
+            compilationStart.RegisterSymbolStartAction(symbolStart =>
+            {
+                var type = (INamedTypeSymbol)symbolStart.Symbol;
+                if (type.IsControllerType())
+                {
+                    symbolStart.RegisterSyntaxNodeAction(nodeContext => ProcessControllerMethods(nodeContext, examinedTypes), SyntaxKind.MethodDeclaration);
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol method
+            && method.IsControllerMethod())
+        {
+            var actionParameters = method.Parameters
+                .Select(x => x.Type)
+                .OfType<INamedTypeSymbol>()
+                .Where(x => !IgnoreType(x))
+                .SelectMany(x => RelatedTypesToExamine(x))
+                .Distinct();
+            foreach (var parameter in actionParameters)
+            {
+                CheckInvalidProperties(parameter, context, examinedTypes);
+            }
+        }
+    }
+
+    private static void CheckInvalidProperties(INamedTypeSymbol parameterType, SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var invalidProperties = GetAllDeclaredProperties(parameterType, examinedTypes)
+                .Where(x => !CanBeNull(x.Type)
+                    && !x.HasAttribute(KnownType.System_ComponentModel_DataAnnotations_RequiredAttribute)
+                    && !x.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ModelBinding_Validation_ValidateNeverAttribute));
+        foreach (var property in invalidProperties)
+        {
+            var propertySyntax = property.DeclaringSyntaxReferences[0].GetSyntax();
+            context.ReportIssue(Rule, propertySyntax.GetLocation());
+        }
+    }
+
+    private static bool IgnoreType(ITypeSymbol type) =>
+        type.IsAny(IgnoredTypes)
+        || (!type.DeclaringSyntaxReferences.Any()
+            && !type.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            && !type.Implements(KnownType.System_Collections_Generic_IEnumerable_T));
+
+    private static bool CanBeNull(ITypeSymbol type) =>
+        type is ITypeParameterSymbol { HasValueTypeConstraint: false }
+        || (type.IsReferenceType && type.NullableAnnotation() != NullableAnnotation.NotAnnotated)
+        || type.IsNullableValueType();
+
+    private static IEnumerable<IPropertySymbol> GetAllDeclaredProperties(ITypeSymbol type, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
```

Comment:
See comment on line 117:
```suggestion
    private static void GetAllDeclaredProperties(ITypeSymbol type, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes, IList<IPropertySymbol> allProperties)
```